In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [6]:
df=pd.read_csv('Data/adult.csv')

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [20]:
df.describe()

age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [21]:
df=df.replace('?',None)

In [22]:
df.isnull().sum().sort_values()

age                  0
fnlwgt               0
education            0
education.num        0
marital.status       0
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
income               0
native.country     583
workclass         1836
occupation        1843
dtype: int64

In [23]:
df.dropna(inplace=True)

In [24]:
df.isnull().sum().sort_values()


age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

In [25]:
df.head()

age workclass  fnlwgt     education  education.num marital.status  \
1   82   Private  132870       HS-grad              9        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   
5   34   Private  216864       HS-grad              9       Divorced   
6   38   Private  150601          10th              6      Separated   

          occupation   relationship   race     sex  capital.gain  \
1    Exec-managerial  Not-in-family  White  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   
5      Other-service      Unmarried  White  Female             0   
6       Adm-clerical      Unmarried  White    Male             0   

   capital.loss  hours.per.week native.country income  
1          4356              18  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K  
5          3770              45  United-States  <=50K  
6          3770              40  United-States  <=50K

In [26]:
df.reset_index(drop=True,inplace=True)

In [27]:
df.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   82   Private  132870       HS-grad              9        Widowed   
1   54   Private  140359       7th-8th              4       Divorced   
2   41   Private  264663  Some-college             10      Separated   
3   34   Private  216864       HS-grad              9       Divorced   
4   38   Private  150601          10th              6      Separated   

          occupation   relationship   race     sex  capital.gain  \
0    Exec-managerial  Not-in-family  White  Female             0   
1  Machine-op-inspct      Unmarried  White  Female             0   
2     Prof-specialty      Own-child  White  Female             0   
3      Other-service      Unmarried  White  Female             0   
4       Adm-clerical      Unmarried  White    Male             0   

   capital.loss  hours.per.week native.country income  
0          4356              18  United-States  <=50K  
1          3900              40  United-States  <=50K  
2          3900              40  United-States  <=50K  
3          3770              45  United-States  <=50K  
4          3770              40  United-States  <=50K

In [28]:
# target = 'income'

# numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns.drop(target, errors='ignore')
# categorical_cols = df.select_dtypes(include=['object']).columns.drop(target)

# total_plots = len(numeric_cols) + len(categorical_cols)
# cols = 3
# rows = (total_plots + cols - 1) // cols

# plt.figure(figsize=(18, rows * 5))

# plot_num = 1

# for col in numeric_cols:
#     plt.subplot(rows, cols, plot_num)
#     sns.boxplot(data=df, x=target, y=col)
#     plt.title(f'{col} vs income')
#     plot_num += 1
#     plt.xticks(rotation = 90)

# for col in categorical_cols:
#     plt.subplot(rows, cols, plot_num)
#     sns.countplot(data=df, x=col, hue=target)
#     plt.xticks(rotation=45)
#     plt.title(f'{col} vs income')
#     plot_num += 1
#     plt.xticks(rotation = 90)

# plt.tight_layout()
# plt.show()

In [29]:
for col in df.columns:
    print(col,df[col].unique(),'\n')

age [82 54 41 34 38 74 68 45 52 32 46 57 37 29 61 51 21 33 49 23 59 60 63 53
 44 43 48 71 73 40 50 42 39 55 47 31 58 62 36 67 72 78 83 26 70 27 35 81
 66 65 25 28 56 69 20 30 90 24 64 75 19 77 22 80 18 17 79 76 88 84 85 86] 

workclass ['Private' 'State-gov' 'Federal-gov' 'Self-emp-not-inc' 'Self-emp-inc'
 'Local-gov' 'Without-pay'] 

fnlwgt [132870 140359 264663 ...  84661 257302 201490] 

education ['HS-grad' '7th-8th' 'Some-college' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-voc' '1st-4th' '5th-6th' 'Assoc-acdm'
 '12th' '9th' 'Preschool'] 

education.num [ 9  4 10  6 16 15 13 14  7 11  2  3 12  8  5  1] 

marital.status ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse'] 

occupation ['Exec-managerial' 'Machine-op-inspct' 'Prof-specialty' 'Other-service'
 'Adm-clerical' 'Transport-moving' 'Sales' 'Craft-repair'
 'Farming-fishing' 'Tech-support' 'Protective-serv' 'Handlers-cleaners'
 'Armed-Fo

In [30]:
df['income']=df['income'].map({'<=50K':0,">50K":1})

In [31]:
from category_encoders import TargetEncoder

ModuleNotFoundError: No module named 'category_encoders'

In [ ]:
target=TargetEncoder(cols=['workclass','education','occupation','native.country'])

In [ ]:
encoding_data=target.fit_transform(df[['workclass','education','occupation','native.country']],df['income'])

In [ ]:
encoding_data

workclass  education  occupation  native.country
0       0.218792   0.164329    0.485220        0.254327
1       0.218792   0.062837    0.124619        0.254327
2       0.218792   0.200060    0.448489        0.254327
3       0.218792   0.164329    0.041096        0.254327
4       0.218792   0.071951    0.133835        0.254327
...          ...        ...         ...             ...
30157   0.218792   0.200060    0.326087        0.254327
30158   0.218792   0.253968    0.304825        0.254327
30159   0.218792   0.164329    0.124619        0.254327
30160   0.218792   0.164329    0.133835        0.254327
30161   0.218792   0.164329    0.133835        0.254327

[30162 rows x 4 columns]

In [ ]:
df[['workclass','education','occupation','native.country']]=encoding_data

In [ ]:
df.head()

age  workclass  fnlwgt  education  education.num marital.status  \
0   82   0.218792  132870   0.164329              9        Widowed   
1   54   0.218792  140359   0.062837              4       Divorced   
2   41   0.218792  264663   0.200060             10      Separated   
3   34   0.218792  216864   0.164329              9       Divorced   
4   38   0.218792  150601   0.071951              6      Separated   

   occupation   relationship   race     sex  capital.gain  capital.loss  \
0    0.485220  Not-in-family  White  Female             0          4356   
1    0.124619      Unmarried  White  Female             0          3900   
2    0.448489      Own-child  White  Female             0          3900   
3    0.041096      Unmarried  White  Female             0          3770   
4    0.133835      Unmarried  White    Male             0          3770   

   hours.per.week  native.country  income  
0              18        0.254327       0  
1              40        0.254327       0  
2              40        0.254327       0  
3              45        0.254327       0  
4              40        0.254327       0

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
le=LabelEncoder()

In [ ]:
df['sex']=le.fit_transform(df['sex'])

In [ ]:
df=pd.get_dummies(df)

In [ ]:
df=df.astype(int)

In [ ]:
df.head()

age  workclass  fnlwgt  education  education.num  occupation  sex  \
0   82          0  132870          0              9           0    0   
1   54          0  140359          0              4           0    0   
2   41          0  264663          0             10           0    0   
3   34          0  216864          0              9           0    0   
4   38          0  150601          0              6           0    1   

   capital.gain  capital.loss  hours.per.week  ...  \
0             0          4356              18  ...   
1             0          3900              40  ...   
2             0          3900              40  ...   
3             0          3770              45  ...   
4             0          3770              40  ...   

   relationship_Not-in-family  relationship_Other-relative  \
0                           1                            0   
1                           0                            0   
2                           0                            0   
3                           0                            0   
4                           0                            0   

   relationship_Own-child  relationship_Unmarried  relationship_Wife  \
0                       0                       0                  0   
1                       0                       1                  0   
2                       1                       0                  0   
3                       0                       1                  0   
4                       0                       1                  0   

   race_Amer-Indian-Eskimo  race_Asian-Pac-Islander  race_Black  race_Other  \
0                        0                        0           0           0   
1                        0                        0           0           0   
2                        0                        0           0           0   
3                        0                        0           0           0   
4                        0                        0           0           0   

   race_White  
0           1  
1           1  
2           1  
3           1  
4           1  

[5 rows x 30 columns]

In [ ]:
df.shape

(30162, 30)

In [ ]:
df=df.drop('capital.loss',axis=1)

In [ ]:


# Q1=df.quantile(0.25)
# Q3=df.quantile(0.75)

# IQR=Q3-Q1
# lower=Q1-1.5*IQR
# upper=Q3+1.5*IQR

# df[numeric_cols]=np.clip(df[numeric_cols],lower,upper,axis=1)

In [ ]:
X=df.drop('income',axis=1)
y=df['income']

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test=train_test_split(X,y,random_state=42,test_size=.2,shuffle=True)

In [ ]:
from sklearn.preprocessing import RobustScaler

In [ ]:
rob=RobustScaler()

In [ ]:
X_train[['age','fnlwgt','hours.per.week','capital.gain']]=rob.fit_transform(X_train[['age','fnlwgt','hours.per.week','capital.gain']])

In [ ]:
X_test[['age','fnlwgt','hours.per.week','capital.gain']]=rob.transform(X_test[['age','fnlwgt','hours.per.week','capital.gain']])

In [ ]:
# from models import run_classification_models,classifications_models


# selected_keys = ['KNN','LogisticRegression']
# selected_models = {}

# for key in selected_keys:
#     selected_models[key] = classifications_models[key]

# result = run_classification_models(selected_models, X_train, X_test, y_train, y_test)

In [ ]:
result

model  train_acc  test_acc  train_f1   test_f1
0                 KNN   0.882921  0.829438  0.880794  0.826135
1  LogisticRegression   0.795143  0.784684  0.806804  0.796873

In [ ]:
# selected_keys = ['SVM']
# selected_models = {}

# for key in selected_keys:
#     selected_models[key] = classifications_models[key]

# result = run_classification_models(selected_models, X_train, X_test, y_train, y_test)

In [ ]:
# result

In [ ]:
import tensorflow as tf


In [ ]:
import tensorflow as tf
from tensorflow import keras

input_dim = X_train.shape[1]

model = Sequential([
    Dense(128, activation='relu', input_shape=(input_dim,)),
    Dropout(0.3),

    Dense(64, activation='relu'),
    Dropout(0.3),

    Dense(32, activation='relu'),

    Dense(1, activation='sigmoid')  # Binary
])

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

history = model.fit(
    X_train, y_train,
    validation_split=0.2,
    epochs=100,
    batch_size=32,
    callbacks=[early_stop],
    verbose=1
)

loss, acc = model.evaluate(X_test, y_test)
print("Test Accuracy:", acc)


2026-02-12 18:46:39.307200: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2026-02-12 18:46:39.384965: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2026-02-12 18:46:39.385026: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2026-02-12 18:46:39.387572: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2026-02-12 18:46:39.399046: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2026-02-12 18:46:39.400013: I tensorflow/core/platform/cpu_feature_guard.cc:1

NameError: name 'X_train' is not defined


model	train_acc	test_acc	train_f1	test_f1
0	KNN	0.882921	0.829438	0.880794	0.826135
2	LogisticRegression	0.795143	0.784684	0.806804	0.796873
3	GaussianNB	0.756061	0.741588	0.771581	0.758251
1	BernoulliNB	0.725227	0.712581	0.743100	0.731592